examples/nlp/multimodal_entailment.py

```diff
@@ -0,0 +1,565 @@
+    preprocessed_2 = keras.applications.resnet_v2.preprocess_input(image_2)
+
+    # Generate the embeddings for the images using the resnet_v2 model
+    # concatenate them.
+    embeddings_1 = resnet_v2(preprocessed_1)
+    embeddings_2 = resnet_v2(preprocessed_2)
+    embeddings = keras.layers.Concatenate()([embeddings_1, embeddings_2])
+
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the vision encoder model.
+    return keras.Model([image_1, image_2], outputs, name="vision_encoder")
+
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers=1, projection_dims=256, dropout_rate=0.1
+)
+keras.utils.plot_model(vision_encoder, show_shapes=True)
+
+
+def create_text_encoder(
+    num_projection_layers, projection_dims, dropout_rate, trainable=False
+):
+    # Load the pre-trained BERT model to be used as the base encoder.
+    bert = hub.KerasLayer(bert_model_path, name="bert",)
+    # Set the trainability of the base encoder.
+    bert.trainable = trainable
+
+    # Receive the text as inputs.
+    bert_input_features = ["input_type_ids", "input_mask", "input_word_ids"]
+    inputs = {
+        feature: keras.Input(shape=(128,), dtype=tf.int32, name=feature)
+        for feature in bert_input_features
+    }
+
+    # Generate embeddings for the preprocessed text using the BERT model.
+    embeddings = bert(inputs)["pooled_output"]
+
+    # Project the embeddings produced by the model.
+    outputs = project_embeddings(
+        embeddings, num_projection_layers, projection_dims, dropout_rate
+    )
+    # Create the text encoder model.
+    return keras.Model(inputs, outputs, name="text_encoder")
+
+
+text_encoder = create_text_encoder(
+    num_projection_layers=1, projection_dims=256, dropout_rate=0.1
+)
+keras.utils.plot_model(text_encoder, show_shapes=True)
+
+
+def create_multimodal_model(
+    num_projection_layers=1,
+    projection_dims=256,
+    dropout_rate=0.1,
+    vision_trainable=False,
+    text_trainable=False,
+):
+    # Receive the images as inputs.
+    image_1 = keras.Input(shape=(128, 128, 3), name="image_1")
+    image_2 = keras.Input(shape=(128, 128, 3), name="image_2")
+
+    # Receive the text as inputs.
+    bert_input_features = ["input_type_ids", "input_mask", "input_word_ids"]
+    text_inputs = {
+        feature: keras.Input(shape=(128,), dtype=tf.int32, name=feature)
+        for feature in bert_input_features
+    }
+
+    # Create the encoders.
+    vision_encoder = create_vision_encoder(
+        num_projection_layers, projection_dims, dropout_rate, vision_trainable
+    )
+    text_encoder = create_text_encoder(
+        num_projection_layers, projection_dims, dropout_rate, text_trainable
+    )
+
+    # Fetch the embedding projections.
+    vision_projections = vision_encoder([image_1, image_2])
+    text_projections = text_encoder(text_inputs)
```

Comment:
@fchollet I want to incorporate a modality dropout trick here to account for the following situation. What if one of the inputs or a pair of the input is not present during inference? How do we make our model perform well under those situations? 

Apparently, modality dropout can be a useful recipe that can help. I am thinking of doing something like the following:

```python
# Fetch the embedding projections.
vision_projections = vision_encoder([image_1, image_2])
vision_projections = keras.layers.Dropout(0.2)(vision_projections)
text_projections = text_encoder(text_inputs)
text_projections = keras.layers.Dropout(0.2)(text_projections)
```

Does this make sense? Or should we introduce `Dropout` at the inputs themselves? I had tried that but it's failing at the text inputs.

Reply:
I had another idea on incorporating cross-attention here:

```python
vision_projections = vision_encoder([image_1, image_2])
text_projections = text_encoder(text_inputs)

# Cross-attention (Luong-style).
query_value_attention_seq = keras.layers.Attention(use_scale=True, dropout=0.2)(
    [vision_projections, text_projections]
)

# Concatenate the projections and pass through the classification layer.
concatenated = keras.layers.Concatenate()([vision_projections, text_projections])
contextual = keras.layers.Concatenate()([concatenated, query_value_attention_seq])
outputs = keras.layers.Dense(3, activation="softmax")(contextual)
```

Would love to get your thoughts on both of the above ideas.

Reply:
Both of these sound fine to me as a way to add regularization. Any particular concern you had?

Reply:
Nothing particular. I just wanted to verify their implementation correctness w.r.t the ideas.